python/dgl/transforms/functional.py

```diff
@@ -1764,13 +1765,23 @@ def remove_nodes(g, nids, ntype=None, store_ids=False):
     g.remove_nodes(nids, ntype=ntype, store_ids=store_ids)
     return g
 
-def add_self_loop(g, etype=None):
+def add_self_loop(g, edge_feat_names=None, fill_data=1., etype=None):
```

Comment:
Generally I prefer a simpler design unless otherwise requested by many users.

- Apply `fill_data` for all edge features.
- Not allow `fill_data` to be a dict.

This will make both the implementation and documentation a lot more cleaner.

Reply:
I am totally agree with you. 

But I would like to discuss with you how many types are allowed in `fill_data`. Right now I have it set to allow `int`, `float`, `str` and `dict`. `float` and `str` are fine in most cases, but `int` is a bit more problematic. If you use aggregation to get edge data, the edge data type is not allowed to be `int`. So I think edge data should always be `float`. As for `dict`, I think it is a shortcut for users to add specific data in a self-loop because `add_node` has a `data` parameter. The `data` parameter is the feature data of the added edge. The key is the feature name and the value is the feature data.

Reply:
> So I think edge data should always be float.

Sometimes edge data are for example binary masks indicating the edges to use for training/validation/test, etc. However, I think this is less an issue as you can always cast the new edge feature to the original data type.

> As for dict, I think it is a shortcut for users to add specific data in a self-loop because add_node has a data parameter. The data parameter is the feature data of the added edge. The key is the feature name and the value is the feature data.

I understand. However this makes things rather complicated when you have multiple edge types and multiple edge features per edge type, particularly when you use `add_self_loop` in `AddSelfLoop`.

Reply:
> Sometimes edge data are for example binary masks indicating the edges to use for training/validation/test, etc. However, I think this is less an issue as you can always cast the new edge feature to the original data type.

For training/validation/testing edge data, it is not appropriate to aggregate the features of the incoming edges of the corresponding nodes, right? But we would like to assume that `fill_data` is applied to all edge features when `edge_feat_names` is `None`. This is actually like the `RowFeatNorm` case at that time. But the case of training/validation/testing edge data is not as much as the case of training/validation/testing node data, so we can leave `edge_feat_names` to the user to make a choice right?

Reply:
It still might make sense if `fill_data` is `1` or `0`. Again it's almost impossible to cover all the cases as likely different edge features and different edge types require different `fill_data`. Let's do a minimal solution first and expand it in the future if this turns out to be really necessary.